Reference/Common-Pitfalls/index.md

```diff
@@ -0,0 +1,466 @@
+
+```csharp
+private static _umbracoContext = UmbracoContext.Current;
+```
+
+```csharp
+//MembershipHelper is also a request scoped instance - it relies either on an UmbracoContext or an HttpContext
+private static _membershipHelper = new MembershipHelper(UmbracoContext.Current);
+```
+
+```csharp
+private static _request = HttpContext.Current.Request;
+```
+
+## Querying with Descendants, DescendantsOrSelf
+
+It's not 100% bad that you use these queries, you just need to understand the implications. 
+Here's a particularly bad scenario:
+
+* You have 10,000 content items in your tree
+* Your tree structure is something like this:
+```
+- Root
+-- Home
+--- Blog (list view with 9495 nodes)
+--- Office Locations (list view with 500 nodes)
+--- About Us
+--- Contact Us
+```
+
+You create a menu on your Home page like:
+
+```
+<ul>
+    <li><a href="@Model.Site().Url">@Model.Site().Name</a></li>
+    @foreach(var node in Model.Site().DescendantsOrSelf().Where(x => x.Level == 2)) 
+    {
+        <li><a href="@node.Url">@node.Name</a></li>
+    }
+</ul>
+```
+
+Which just renders out: _Home, Blog, Office Locations, About Us, Contact Us_
+
+BUT! ...  this is going to perform most horribly. This is going to iterate over every single node in Umbraco, all 10,000 of them. Further more, 
+this means it is going to allocate 10,000 `IPublishedContent` instances in memory just in order to check it's `Level` value. 
+
+This can easily be re-written as:
+
+```
+<ul>
+    <li><a href="@Model.Site().Url">@Model.Site().Name</a></li>
+    @foreach(var node in Model.Site().Children) 
+    {
+        <li><a href="@node.Url">@node.Name</a></li>
+    }
+</ul>
+```
+
+In many cases you might know that there is only ever going to be a small number of Descendants and if so then go nuts and use Descendants or DescendantsOrSelf, 
+it's just important to be aware of the implications of what you are writing.  
+
+## Too much querying (Over querying)
+
+Querying content is not Free! Anytime you make a query or resolve a property value be aware that there is overhead involved. 
+You could try to think about every query you make as an SQL call - you don't want to make too many otherwise the performance of 
+your website is going to suffer.
+
+Here's a common pitfall that is seen. Let's continue the menu example, in this example the menu is going to be rendered
+using the current page's root node::
+
+```
+<ul>
+    <li><a href="@Model.Site().Url">@Model.Site().Name</a></li>
+    @foreach(var node in Model.Site().Children) 
+    {
+        <li><a href="@node.Url">@node.Name</a></li>
+    }
+</ul>
+```
+
+The syntax `@Model.Content.Site()` is actually shorthand for doing this:
+`Model.Content.AncestorsOrSelf(1)` which means it is going to traverse up the tree until it reaches an ancestor node
+with a level of one. As mentioned above, traversing costs resources and in this example there is 3x traversals being done
+for the same value. Instead this can be rewritten as:
+
+```
+@{
```

Comment:
Again checked the query builder breaks it out into variable first before iterating over it & it does like so. So all good again :)
```
@{
	var selection = CurrentPage.Site().Children.Where("Visible");
}
<ul>
	@foreach(var item in selection){
		<li>
			<a href="@item.Url">@item.Name</a>
		</li>
	}
</ul>
```